### PR TITLE
Updated .eslintrc and fixed a whole bunch of linting errors and warnings

### PR DESCRIPTION
**Pull request by greg-js**
So I noticed quite a bit of linting errors and warnings so I went through them, fixed the obvious ones, addressed some less obvious ones and updated the eslint config files for those cases where the rules didn't seem to reflect the style choices used in the project.

I tried to stick as close as possible to the style used and I'm pretty sure I didn't break anything. But hey, with changes like these, accidental breakage _is_ possible so you should probably go through the changes before accepting the PR. If I don't fall asleep later I'll go through it myself and add github comments wherever a certain change may seem controversial.

Even with these changes though, `npm run lint` still finds three errors and a warning. One of them (the one in `./app/reducers/keychainReducer`) I wasn't sure how to handle. The others I'm assuming are a result of as of yet unfinished features?

Also, I understand that stylistic changes can be very personal and controversial. It's by no means my intention to enforce any kind of style on anyone -- honestly, if it were up to me, there would be about 1000000 more semicolons in this project. I just want to help you get closer to a perfect state of no-linting-errors development bliss. If you don't like the spirit of this PR, just close it, I have no feelings so they can't get hurt.

Oh, and can you tell that a) I really like the idea behind this project, and b) I'm currently funemployed? ;-)

P.S. I realize I've probably been bombarding your inbox by adding comments on the changes but I regret nothing

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -14,6 +24,8 @@ module.exports = {
     'constructor-super': 2,
     'quote-props': [2, 'consistent'],
     'no-underscore-dangle': [1],
-    'semi': [2, 'never']
+    'semi': [2, 'never'],
+    'no-console': [0],
+    'new-cap': [1, { newIsCap: true, capIsNew: false }],
```
Comment (greg-js) [inline .eslintrc.js]: http://eslint.org/docs/rules/new-cap

File: app/components/composer/ComposerForm.js [hunk 2]
```diff
@@ -109,15 +109,23 @@ class ComposerForm extends Component {
...
+  doPgp = (pgpFn, ...options) => {
     return () => {
       const data = this.refs.textarea.value
       const output = pgpFn(data, ...options)
       return output
     }
   }
 
-  readKey = this._pgp(readArmored)
+  readKey = this.doPgp(readArmored)
```
Comment (greg-js) [inline app/components/composer/ComposerForm.js]: alternatively, you could change the `no-underscore-dangle` rule

File: app/components/floating-button/FloatingButton.js [hunk 3]
```diff
@@ -37,6 +37,10 @@ class FloatingButton extends Component {
 
   handleClose = () => this.setState({ isShowingActions: false })
 
+  showComposer(type) {
+    return this.props.showComposer.bind(this, type)
```
Comment (greg-js) [inline app/components/floating-button/FloatingButton.js]: This seems like a pretty elegant solution to me, but afaik you'd still be getting the performance hit from creating a new function on the fly every time, which is what the eslint rule was trying to prevent.
I can't immediately think of a way around it though because we can't bind a method on `this.props` in the constructor..

File: app/components/header/HeaderKeyStatusSpinner.js [hunk 4]
```diff
@@ -1,21 +1,18 @@
...
-class HeaderKeyStatusSpinner extends Component {
-
-  render() {
-    return (
-      <div>
-        <div className="spinner">
-          <div className="double-bounce1"></div>
-          <div className="double-bounce2"></div>
-        </div>
+const HeaderKeyStatusSpinner = () => {
```
Comment (greg-js) [inline app/components/header/HeaderKeyStatusSpinner.js]: the existing eslint config prefers stateless functional components and it does seem to make sense in this case, of course you could just change the rule if you don't like it this way

File: app/utils/icons.js [hunk 5]
```diff
@@ -22,9 +22,9 @@ const store = {
   'plus': `<svg style="width:100%;height:100%" viewBox="0 0 24 24">
     <path fill="currentColor" d="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z" />
   </svg>`, // materialdesignicons.com plus
-  'check': `<svg style="width:24px;height:24px" viewBox="0 0 24 24">
-    <path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z" />
-  </svg>`,
+  // 'check': `<svg style="width:24px;height:24px" viewBox="0 0 24 24">
```
Comment (greg-js) [inline app/utils/icons.js]: I wasn't sure if maybe you had a reason for this so I just commented out the duplicate `check` property rather than deleting it

File: app/actions/AliasesActions.js [hunk 6]
```diff
@@ -1,6 +1,6 @@
 import * as types from '../constants/AliasesConstants'
 import db from '../config/database.js'
-import { generateKey } from '../../utils/pgp'
+import { generateKey } from '../utils/pgp'
```
Comment (greg-js) [inline app/actions/AliasesActions.js]: actually, it seems this whole file is not actually used and has broken imports. I assume it's because it is/was still in development

File: app/components/floating-button/FloatingButton.js (hunk 3 above)
Comment (tobycyanide) [inline app/components/floating-button/FloatingButton.js]: Can't you do something like: 
```js
constructor(props) {
    super(props);
    this.showComposer = ::this.props.showComposer
}
```
the `::` is the equivalent of `.bind(this)`, and the `type` arg is gonna be passed anyway, this is normally what I do in my react projects.

Comment (greg-js) [inline app/components/floating-button/FloatingButton.js]: Oh! You're right! It's very very late and for some reason I was thinking about reassigning the method on `this.props` instead of just on `this`. I'll change it. I'll stick to just using `bind` though as more people will immediately know what's going on. Thanks for the heads up :-)

**Comment by greg-js** [comment]
Well it seems that somewhere along the way, an error crept in and the composer now pops up when hovering over the floating button. I need sleep though so I'll check what went wrong in the morning, but do not merge yet! :-)

**Comment by greg-js** [comment]
So in the end putting the binding in the constructor caused issues, and using an arrow function in render left the linting error, so I added a workaround by factoring out the floating button div (and rolled back the change in the eslint sorting order). So the bug is fixed and now down to three linting problems.

**Comment by henryboldi** [comment]
Excellent! Thanks for posting updates on your progress 😃

File: app/utils/pgp.js [hunk 7]
```diff
@@ -1,7 +1,9 @@
 const openpgp = require('openpgp')
 openpgp.initWorker({ path: '../node_modules/openpgp/dist/openpgp.worker.min.js' })
+/* eslint-disable */
```
Comment (casesandberg) [inline app/utils/pgp.js]: Can you just disable each of the lines individually instead of using this block disable?

File: app/utils/icons.js (hunk 5 above)
Comment (casesandberg) [inline app/utils/icons.js]: Just delete it, thanks

File: .eslintrc.js [hunk 8]
```diff
@@ -5,6 +5,16 @@ module.exports = {
     // "jsx-curly-spacing": [2, "always", {"alternative": true}], // https://github.com/yannickcr/eslint-plugin-react/issues/388
     'react/jsx-curly-spacing': [1, 'always'],
     'react/prop-types': [0],
+    'react/sort-comp': [1, {
```
Comment (casesandberg) [inline .eslintrc.js]: Can we not do something custom here and just use the default sort order?

File: app/actions/AliasesActions.js [hunk 9]
```diff
@@ -1,15 +1,15 @@
...
 export function addAlias(alias) {
-  return async function (dispatch) {
+  return async function dispatchAction(dispatch) {
     try {
       const generatedKey = await generateKey()
       const insertedAlias = await db('aliases').insert(alias)
       return dispatch({ type: types.ADD_KEY, alias: insertedAlias })
     } catch (err) {
-      console.log(err)
+      return console.log(err)
```
Comment (casesandberg) [inline app/actions/AliasesActions.js]: Never return a console.log

File: app/actions/AliasesActions.js [hunk 10]
```diff
@@ -19,12 +19,12 @@ export function setAliases(aliases) {
...
 
 export function fetchAliases() {
-  return async function (dispatch) {
+  return async function dispatchAction(dispatch) {
     try {
       const aliases = await db('aliases').value()
       return dispatch(setAliases(aliases))
     } catch (err) {
-
+      return console.log(err)
```
Comment (casesandberg) [inline app/actions/AliasesActions.js]: Never return a console.log

File: app/actions/KeychainActions.js [hunk 11]
```diff
@@ -2,12 +2,12 @@ import * as types from '../constants/KeychainConstants'
...
 
 export function addKey(key) {
-  return async function (dispatch) {
+  return async function dispatchAction(dispatch) {
     try {
       const insertedKey = await db('keychain').insert(key)
       return dispatch({ type: types.ADD_KEY, key: insertedKey })
     } catch (err) {
-      console.log(err)
+      return console.log(err)
```
Comment (casesandberg) [inline app/actions/KeychainActions.js]: Never return a console.log

File: app/actions/KeychainActions.js [hunk 12]
```diff
@@ -17,12 +17,12 @@ export function setKeychain(keychain) {
...
 
 export function fetchKeychain() {
-  return async function (dispatch) {
+  return async function dispatchAction(dispatch) {
     try {
       const keychain = await db('keychain').value()
       return dispatch(setKeychain(keychain))
     } catch (err) {
-
+      return console.log(err)
```
Comment (casesandberg) [inline app/actions/KeychainActions.js]: Never Return a console.log

File: app/components/floating-button/FloatingButton.js [hunk 13]
```diff
@@ -37,6 +37,19 @@ class FloatingButton extends Component {
 
   handleClose = () => this.setState({ isShowingActions: false })
 
+  renderFloatingButtonDiv(type) {
```
Comment (casesandberg) [inline app/components/floating-button/FloatingButton.js]: Can you pull this change out into another PR? I really dig it.

File: app/components/header/HeaderKeyStatusSpinner.js [hunk 14]
```diff
@@ -1,21 +1,18 @@
 'use strict'
 
-import React, { Component } from 'react'
+import React from 'react'
 
-import colors from '../../assets/styles/variables/colors'
+// import colors from '../../assets/styles/variables/colors'
```
Comment (casesandberg) [inline app/components/header/HeaderKeyStatusSpinner.js]: Delete

File: app/components/header/HeaderKeyStatusTooltip.js [hunk 15]
```diff
@@ -4,7 +4,7 @@ import React, { Component } from 'react'
 import ReactCSS from 'reactcss'
 import dynamics from 'dynamics.js'
 
-import { Icon } from '../common/index'
+// import { Icon } from '../common/index'
```
Comment (casesandberg) [inline app/components/header/HeaderKeyStatusTooltip.js]: Delete

File: app/containers/App.js [hunk 16]
```diff
@@ -1,23 +1,23 @@
...
-export default class App extends Component {
-  static propTypes = {
-    children: PropTypes.element.isRequired,
-  }
+const App = (props) => {
+  return (
+    <div>
+      { props.children }
+      {
+        (() => {
```
Comment (casesandberg) [inline app/containers/App.js]: This looks weird. Maybe assign the function in the top of the function and point to it here?

File: app/containers/ComposerContainer.js [hunk 17]
```diff
@@ -1,6 +1,10 @@
 import _ from 'lodash'
 import { connect } from 'react-redux'
-import { toggleComposer, clearSelectedKeys, toggleGeneratingKey, addKey, setOutput } from '../actions/index'
+import { toggleComposer,
+         // clearSelectedKeys,
```
Comment (casesandberg) [inline app/containers/ComposerContainer.js]: Delete

File: app/containers/ComposerContainer.js [hunk 18]
```diff
@@ -1,6 +1,10 @@
 import _ from 'lodash'
 import { connect } from 'react-redux'
-import { toggleComposer, clearSelectedKeys, toggleGeneratingKey, addKey, setOutput } from '../actions/index'
+import { toggleComposer,
```
Comment (casesandberg) [inline app/containers/ComposerContainer.js]: Put this back to one line.

File: app/reducers/keychainReducer.js [hunk 19]
```diff
@@ -31,7 +31,7 @@ const keychainReducer = (state = initialState, action) => {
       )
     case SET_KEYCHAIN:
       return state.withMutations((st) => {
-        for (let key of action.keychain) {
+        for (const key of action.keychain) {
```
Comment (casesandberg) [inline app/reducers/keychainReducer.js]: Are you sure this works without reassign issues?

**Comment by greg-js** [comment]
Cool, thank you for the pointers, will use your comments to update the PR later today or tomorrow.

File: .eslintrc.js (hunk 8 above)
Comment (greg-js) [inline .eslintrc.js]: I don't mind putting the nine components that have sort order lint errors in the default sort order. It would have some awkward effects though, particularly it would result in the `classes` method from `ReactCSS` being counted as a user-defined method (and thus placed in between the life cycle methods and `render`). I will just take out the custom sorting rule for now.

File: app/actions/KeychainActions.js (hunk 12 above)
Comment (greg-js) [inline app/actions/KeychainActions.js]: Sorry, did not realize it was not-done. How about returning `void 0`?

File: app/components/floating-button/FloatingButton.js (hunk 13 above)
Comment (greg-js) [inline app/components/floating-button/FloatingButton.js]: Will do.

File: app/containers/App.js (hunk 16 above)
Comment (greg-js) [inline app/containers/App.js]: This one I'd like to test against a current version before I submit changes because I'm currently not getting Redux DevTools to work at all. Sadly I am running into #81 now on a fresh install (it's the electron asar module version mismatch thing all over again - I'm on Linux) so I'll revert back to master.

FWIW, [here's the recommended setup](https://github.com/gaearon/redux-devtools/blob/master/docs/Walkthrough.md#containersrootjs) for DevTools, would be nice and clean IMO if it works

File: app/reducers/keychainReducer.js (hunk 19 above)
Comment (greg-js) [inline app/reducers/keychainReducer.js]: As long as `key` isn't modified, it works fine. It's in the spec and also specifically noted as an example [on MDN](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Statements/for...of). I did do a little digging and saw it's not well implemented in Firefox, but not relevant here.

I agree at first sight it looks absurd, but it does fix a linting issue and it's also kind of cool in a way.. ;)

**Comment by greg-js** [comment]
Uhh, I'm not sure what happened there, it seemed fine on my end until I pushed it. Well, I could create a new, clean PR later with these linter-pleasing changes if you'd like. I'll wait until you've looked at it though.

Remaining linting issues are:
* in the two unused files (`app/actions/AliasesActions.js` & `./app/actions/KeychainActions.js`)
* in some unused (but maybe somehow relevant for later?) code in `./app/components/composer/ComposerAliasForm.js` (line 147)
* in `./app/containers/App.js`, see my comment above
* about the `classes` method being in the wrong place according to the sort order
* about the floating button for which I'll make another pull request soon

**Comment by casesandberg** [comment]
Thanks @greg-js, I plucked out your changes and merged them into master!
